crates/ra_ide/src/diagnostics.rs

```diff
@@ -711,4 +729,261 @@ fn main() {
             check_struct_shorthand_initialization,
         );
     }
+
+    #[test]
+    fn test_match_arms_empty_tuple_no_arms() {
```

Comment:
Just a reminder that these detailed tests for the diagnostic should live in hir_ty, not here. That would probably also reduce the time you have to wait for builds to finish before you can run tests ;)

Reply:
Faster builds would be great. I've just started kindof a big change to get enums working, so I think I'll make this move after that. Thanks for the suggestion!

Reply:
I'm at a point now where I could move these tests, but upon further review I realized that I think this is something we worked through previously in discussion in this PR and determined perhaps a refactor to move the tests could follow this PR? As far as I can tell I am following the pattern set by the missing fields diagnostic. Am I missing some context here?

Reply:
We already have a test for the missing fields diagnostic in hir_ty [here](https://github.com/rust-analyzer/rust-analyzer/blob/6207ac90da7e607218457bf872fa32df6b78f224/crates/ra_hir_ty/src/tests.rs#L299). That's where I'm suggesting the tests for this diagnostic should be for now as well :slightly_smiling_face: The tests in ra_ide are supposed to test the fixes for the diagnostic, which I think are supposed to stay in ra_ide; some of the 'no diagnostic' tests are ones that probably should live in hir_ty though. 

So I think the 'end state' that we want is that the implementation of the diagnostic is in one module in hir_ty, together with a bunch of tests checking cases where the diagnostic should and should not appear; and then ra_ide implements a quick fix for the diagnostic, together with tests checking specifically the fix. (Correct me if I'm wrong here @matklad). 

But for now putting more tests in hir_ty should already be possible.

Reply:
Thanks for spelling that out for me. I've moved the tests to the `_match` module. 

I also made a minor change to the `TestDB::Diagnostics` method, to allow it to return a diagnostic count as well as the (existing) diagnostic message buffer. My rationale here is that we should probably have a relatively small quantity of tests that assert that the diagnostic message is correct, and a larger quantity of tests who only check that a diagnostic was or was not triggered. I think this significantly reduces the effort required to create/maintain the tests. Let me know if you see any issues with this approach.

Reply:
Sounds plausible! For testing the message I think it'd be a good idea to continue using snapshots though.

Reply:
> For testing the message I think it'd be a good idea to continue using snapshots though.

That works for me. I've switched back to using snapshot for the diagnostic message checks.